vms/platformvm/txs/executor/proposal_tx_executor.go

```diff
@@ -320,215 +322,303 @@ func (e *ProposalTxExecutor) RewardValidatorTx(tx *txs.RewardValidatorTx) error
+			stakerToReward.TxID,
 			tx.TxID,
 		)
 	}
 
 	// Verify that the chain's timestamp is the validator's end time
 	currentChainTime := e.OnCommitState.GetTimestamp()
-	if !stakerToRemove.EndTime.Equal(currentChainTime) {
+	if !stakerToReward.NextTime.Equal(currentChainTime) {
 		return fmt.Errorf(
 			"%w: TxID = %s with %s < %s",
 			ErrRemoveStakerTooEarly,
 			tx.TxID,
 			currentChainTime,
-			stakerToRemove.EndTime,
+			stakerToReward.NextTime,
 		)
 	}
 
+	// retrieve primaryNetworkValidator before possibly removing it.
 	primaryNetworkValidator, err := e.OnCommitState.GetCurrentValidator(
 		constants.PrimaryNetworkID,
-		stakerToRemove.NodeID,
+		stakerToReward.NodeID,
 	)
 	if err != nil {
 		// This should never error because the staker set is in memory and
 		// primary network validators are removed last.
 		return err
 	}
 
-	stakerTx, _, err := e.OnCommitState.GetTx(stakerToRemove.TxID)
+	stakerTx, _, err := e.OnCommitState.GetTx(stakerToReward.TxID)
 	if err != nil {
 		return fmt.Errorf("failed to get next removed staker tx: %w", err)
 	}
 
+	// Invariant: A [txs.DelegatorTx] does not also implement the
+	//            [txs.ValidatorTx] interface.
 	switch uStakerTx := stakerTx.Unsigned.(type) {
 	case txs.ValidatorTx:
-		e.OnCommitState.DeleteCurrentValidator(stakerToRemove)
-		e.OnAbortState.DeleteCurrentValidator(stakerToRemove)
+		if err := e.rewardValidatorTx(uStakerTx, stakerToReward); err != nil {
+			return err
+		}
+	case txs.DelegatorTx:
+		if err := e.rewardDelegatorTx(uStakerTx, stakerToReward); err != nil {
+			return err
+		}
+	default:
+		// Invariant: Permissioned stakers are removed by the advancement of
+		//            time and the current chain timestamp is == this staker's
+		//            EndTime. This means only permissionless stakers should be
+		//            left in the staker set.
+		return ErrShouldBePermissionlessStaker
+	}
+
+	// If the reward is aborted, then the current supply should be decreased.
+	currentSupply, err := e.OnAbortState.GetCurrentSupply(stakerToReward.SubnetID)
+	if err != nil {
+		return err
+	}
+	newSupply, err := safemath.Sub(currentSupply, stakerToReward.PotentialReward)
+	if err != nil {
+		return err
+	}
+	e.OnAbortState.SetCurrentSupply(stakerToReward.SubnetID, newSupply)
+
+	// here both commit and abort state supplies are correct. Handle staker lifecycle.
+	if _, ok := stakerTx.Unsigned.(txs.ValidatorTx); ok {
+		e.OnCommitState.DeleteCurrentValidator(stakerToReward)
+		e.OnAbortState.DeleteCurrentValidator(stakerToReward)
+	} else { // must be txs.DelegatorTx due to switch check above
+		e.OnCommitState.DeleteCurrentDelegator(stakerToReward)
+		e.OnAbortState.DeleteCurrentDelegator(stakerToReward)
+	}
 
-		stake := uStakerTx.Stake()
-		outputs := uStakerTx.Outputs()
+	// handle option preference
+	shouldCommit, err := e.calculateProposalPreference(stakerToReward, primaryNetworkValidator)
+	if err != nil {
+		return err
+	}
+
+	e.PrefersCommit = shouldCommit
```

Comment:
This is the main cleanup. RewardValidatorTx processing is broken into three pieces:
1. Calculate reward
2. Duly update current supply
3. handle stakers lifetime (just removing them here)
4. handle node preferences (should staker be rewarded or not?)